api/src/test/java/org/openmrs/api/ProgramWorkflowServiceUnitTest.java

```diff
@@ -173,4 +179,146 @@ public void saveProgram_shouldFailIfProgramWorkFlowStateTerminalIsNull() {
 		
 		pws.saveProgram(program);
 	}
+	
+	@Test
+	public void savePatientProgram_shouldFailForNullPatient() {
+		
+		exception.expect(APIException.class);
+		exception.expectMessage("PatientProgram requires a Patient and a Program");
+		
+		PatientProgram patientProgram = new PatientProgram(1);
+		patientProgram.setProgram(new Program(1));
+		
+		pws.savePatientProgram(patientProgram);
+	}
+	
+	@Test
+	public void savePatientProgram_shouldFailForNullProgram() {
+		
+		exception.expect(APIException.class);
+		exception.expectMessage("PatientProgram requires a Patient and a Program");
+		
+		PatientProgram patientProgram = new PatientProgram(1);
+		patientProgram.setPatient(new Patient(1));
+		
+		pws.savePatientProgram(patientProgram);
+	}
+	
+	@Test
+	public void purgePatientProgram_shouldFailForNotEmptyStatesAndTrueCascade() {
+		
+		exception.expect(APIException.class);
+		exception.expectMessage("Cascade purging of PatientPrograms is not implemented yet");
+		
+		PatientProgram patientProgram = new PatientProgram(1);
+		
+		Set<PatientState> patientStates = new Set<PatientState>() {
+			
+			private ArrayList<PatientState> states = new ArrayList();
+			
+			@Override
+			public int size() {
+				return 0;
+			}
+			
+			@Override
+			public boolean isEmpty() {
+				return false;
+			}
+			
+			@Override
+			public boolean contains(Object o) {
+				return false;
+			}
+			
+			@Override
+			public Iterator<PatientState> iterator() {
+				return null;
+			}
+			
+			@Override
+			public Object[] toArray() {
+				return new Object[0];
+			}
+			
+			@Override
+			public <T> T[] toArray(T[] a) {
+				return null;
+			}
+			
+			@Override
+			public boolean add(PatientState patientState) {
+				states.add(patientState);
+				return true;
+			}
+			
+			@Override
+			public boolean remove(Object o) {
+				return false;
+			}
+			
+			@Override
+			public boolean containsAll(Collection<?> c) {
+				return false;
+			}
+			
+			@Override
+			public boolean addAll(Collection<? extends PatientState> c) {
+				return false;
+			}
+			
+			@Override
+			public boolean retainAll(Collection<?> c) {
+				return false;
+			}
+			
+			@Override
+			public boolean removeAll(Collection<?> c) {
+				return false;
+			}
+			
+			@Override
+			public void clear() {
+				//doSomething
+				
+			}
+			
+			@Override
+			public boolean equals(Object o) {
+				return false;
+			}
+			
+			@Override
+			public int hashCode() {
+				return 0;
+			}
+		};
+		
+		PatientState patientState = new PatientState();
+		patientState.setState(new ProgramWorkflowState(1));
+		
+		patientStates.add(patientState);
+		
+		patientProgram.setStates(patientStates);
+		
+		
+		pws.purgePatientProgram(patientProgram, true);
+		
+	}
+	
+	@Test
+	public void purgeProgram_shouldFailForNotEmptyWorkFlowsAndTrueCascade() {
+		
+		exception.expect(APIException.class);
+		exception.expectMessage("Cascade purging of Programs is not implemented yet");
+		
+		Program program = new Program(1);
+		
```

Comment:
remove newlines in between your setup. Remember the arrange/act/assert. Grouping related code together makes it more readable.

```java
  ...
  exception.expectMessage("Cascade purging of Programs is not implemented yet");

  Program program = new Program(1);
  ProgramWorkflow workflow = new ProgramWorkflow(1);
  program.addWorkflow(workflow);

  pws.purgeProgram(program, true);
}
```

Reply:
@teleivo I am sorry for keep forgetting the basics